server/src/main/java/org/apache/druid/metadata/input/SqlEntity.java

```diff
@@ -0,0 +1,198 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.metadata.input;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Preconditions;
+import org.apache.druid.data.input.InputEntity;
+import org.apache.druid.java.util.common.StringUtils;
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.metadata.SQLFirehoseDatabaseConnector;
+import org.skife.jdbi.v2.ResultIterator;
+import org.skife.jdbi.v2.exceptions.CallbackFailedException;
+import org.skife.jdbi.v2.exceptions.ResultSetException;
+import org.skife.jdbi.v2.exceptions.StatementException;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Represents a rdbms based input resource and knows how to read query results from the resource using SQL queries.
+ */
+public class SqlEntity implements InputEntity
+{
+  private static final Logger LOG = new Logger(SqlEntity.class);
+
+  private final String sql;
+  private final ObjectMapper objectMapper;
+  private final SQLFirehoseDatabaseConnector sqlFirehoseDatabaseConnector;
+  private final boolean foldCase;
+
+  public SqlEntity(
+      String sql,
+      SQLFirehoseDatabaseConnector sqlFirehoseDatabaseConnector,
+      boolean foldCase,
+      ObjectMapper objectMapper
+  )
+  {
+    this.sql = sql;
+    this.sqlFirehoseDatabaseConnector = Preconditions.checkNotNull(
+        sqlFirehoseDatabaseConnector,
+        "SQL Metadata Connector not configured!"
+    );
+    this.foldCase = foldCase;
+    this.objectMapper = objectMapper;
+  }
+
+  public String getSql()
+  {
+    return sql;
+  }
+
+  @Nullable
+  @Override
+  public URI getUri()
+  {
+    return null;
+  }
+
+  @Override
+  public InputStream open()
+  {
+    throw new UnsupportedOperationException("Please use fetch() instead");
+  }
+
+  @Override
+  public CleanableFile fetch(File temporaryDirectory, byte[] fetchBuffer) throws IOException
+  {
+    final File tempFile = File.createTempFile("druid-sql-entity", ".tmp", temporaryDirectory);
+    return openCleanableFile(sql, sqlFirehoseDatabaseConnector, objectMapper, foldCase, tempFile);
+
+  }
+
+  public static CleanableFile openCleanableFile(
+      String sql,
+      SQLFirehoseDatabaseConnector sqlFirehoseDatabaseConnector,
+      ObjectMapper objectMapper,
+      boolean foldCase,
+      File tempFile
+  )
+      throws IOException
+  {
+    try (FileOutputStream fos = new FileOutputStream(tempFile)) {
+      final JsonGenerator jg = objectMapper.getFactory().createGenerator(fos);
```

Comment:
Please move `jg` to the above `try-resource` clause so that i can be closed safely.